### PR TITLE
feat(snowflake): warn on password-auth deprecation in CLI

### PR DESCRIPTION
**Pull request by alokr-dhub**
## Summary

Surfaces a deprecation warning when a Snowflake ingestion recipe is configured with username + password authentication (`DEFAULT_AUTHENTICATOR`).

Snowflake is deprecating username + password authentication as part of its Strong Authentication rollout. Password auth will stop working on an account-specific enforcement date during Phase 3 (Aug–Oct 2026). This change warns operators at recipe load time so they can migrate to key-pair auth before their account's enforcement date.

## What's included

- **`snowflake_connection.py`**: `SnowflakeConnectionConfig.validate_authentication_config` now emits the deprecation when `authentication_type` is `DEFAULT_AUTHENTICATOR` (or unset) **and** a non-empty password is present. It surfaces through two channels:
  - `add_global_warning` — reaches the CLI summary's `pretty_print_summary`, so every consumer of `SnowflakeConnectionConfig` (Snowflake sources, Fivetran destinations, the tag propagator) sees it once.
  - `warnings.warn(..., ConfigurationWarning)` — reaches Python's warnings system, which is the only channel `--test-source-connection` prints (that path logs only the connection report JSON, never the summary).
  - The warning text and migration-guide URL live in one module-level helper (`get_password_auth_deprecation_warning()`) so CLI and report messages can't drift.
- **`snowflake_v2.py`**: `SnowflakeV2Source.__init__` mirrors the deprecation via `self.report.warning(...)` so it also surfaces in the structured ingestion report / DataHub UI (global warnings never reach the source report).
- **Unit tests** in `test_snowflake_source.py` covering the `bool(SecretStr(""))` edge case (the wrapper is always truthy, so `get_secret_value()` must be read first) and the `ConfigurationWarning` emission on validate.

## Behavior

Soft warning only — existing recipes keep running, with a clear migration message pointing to key-pair auth. No hard-fail toggle; Snowflake enforces the cutoff on each account's enforcement date.

## Test plan

- [x] `./gradlew :metadata-ingestion:lintFix` passes (ruff check + ruff format)
- [x] `./gradlew :metadata-ingestion:lint` passes (ruff + mypy)
- [x] `pytest tests/unit/snowflake/test_snowflake_source.py` — 79 passed
- [ ] Manual: run a password-auth Snowflake recipe and confirm the warning appears in CLI output and the ingestion report

---
## Summary by cubic
Warns operators when a Snowflake ingestion recipe uses username/password auth, since Snowflake deprecates it during its Strong Authentication rollout (Phase 3, Aug–Oct 2026). Recipes previously ran with no warning; now config validation emits a soft warning, with no hard-fail toggle — Snowflake enforces the cutoff on each account's enforcement date.

- Detects password auth via `is_using_password_auth()` on `SnowflakeConnectionConfig`, treating an empty password as unset.
- Surfaces the warning via `warnings.warn(ConfigurationWarning)` for `--test-source-connection` and via `SnowflakeV2Source`'s structured report for the DataHub UI.
- Keeps the warning text and migration-guide URL as shared constants in `snowflake_connection.py` so the two channels can't drift.

<sup>Written for commit 747ea4dfd70f68f0413acded1300f03b42a29929. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/datahub-project/datahub/pull/19693?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Comment by alokr-dhub** [comment]
Addressed all 5 cubic findings in c5194db:

1. **Duplicate test (P3)** — Replaced `test_is_using_password_auth_flags_unset_type_with_password` (which was a verbatim duplicate of the test above it) with `test_is_using_password_auth_ignores_empty_secretstr`, which tests the empty-`SecretStr` edge case. The unset-`authentication_type` scenario is already covered by `test_config_validation_adds_global_warning_for_password_auth` (omits the field, relies on the pydantic default).

2. **`bool(SecretStr)` (P2)** — Added a `_secret_value()` helper that calls `get_secret_value()` on the password before applying truthiness, so an empty `SecretStr` (whose wrapper is always truthy) is no longer mistaken for a populated password.

3. **`--test-source-connection` doesn't print global warnings (P2)** — Verified: `_test_source_connection` only logs the connection report JSON; `Pipeline.pretty_print_summary` (which prints global warnings) is never called on that path. Added a `logger.warning(get_password_auth_deprecation_warning())` in `SnowflakeV2Source.test_connection` so the deprecation surfaces in that CLI output.

4. **Divergent message (P2)** — `SnowflakeV2Source.__init__` now calls `get_password_auth_deprecation_warning()` instead of re-typing a divergent message. The CLI global warning and the structured-report warning share one source of truth.

5. **Warning after `get_connection()` (P2)** — Moved the `self.report.warning` call before `self.config.get_connection()` so the deprecation still reaches the report even if Snowflake rejects password auth during connection creation.

Verification: `ruff check` + `ruff format --check` + `mypy` (no issues in 2727 files) pass; 13 deprecation tests pass; 868 snowflake unit tests pass (no regressions).

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_v2.py [hunk 1]
```diff
@@ -177,6 +180,14 @@ def __init__(self, ctx: PipelineContext, config: SnowflakeV2Config):
         # The exit stack helps ensure that we close all the resources we open.
         self._exit_stack = contextlib.ExitStack()
 
+        # Before get_connection() so the warning still reaches the report if
+        # Snowflake rejects password auth during the connect call.
+        if self.config.is_using_password_auth():
+            self.report.warning(
+                get_password_auth_deprecation_warning(),
```
[the remark this answers is not in the document]
Comment (alokr-dhub) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_v2.py]: Deduplicated in `98732ff`: dropped the shared `add_global_warning` from the config validator (kept only the hard-error raise) and now surface the soft warning through each Snowflake source's own report — the standard pattern every other source uses. `datahub ingest run` prints the deprecation exactly once (Source report section), the structured report still carries it for the UI, and the summary/queries sources get the same report warning so they keep coverage. The `--test-source-connection` path already logs directly.

[the remark this answers is not in the document]
Comment (alokr-dhub) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_v2.py]: Deduplicated in `e462c307`: restored `add_global_warning` in the shared config validator as the single CLI channel (covers every consumer of SnowflakeConnectionConfig — Snowflake sources, Fivetran destinations, the tag propagator) and removed the duplicate `self.report.warning` from SnowflakeV2Source and SnowflakeSummarySource. `datahub ingest run` now prints the deprecation exactly once. The `--test-source-connection` path still logs directly.

**Review by treff7es — CHANGES_REQUESTED** [review]
A few notes on the shape of this rather than the intent — surfacing the deprecation is clearly worth doing.

**Most of `snowflake_auth_deprecation.py` already exists.** `datahub/configuration/validate_field_deprecation.py` has `pydantic_field_deprecated`, used by ~10 connectors (dbt, tableau, looker, powerbi, s3, …). It handles the `add_global_warning` + `warnings.warn(..., ConfigurationWarning)` pair and the validator-name dedup hack pydantic needs. This looks like it collapses to one line in the class body:

```python
_password_auth_deprecated = pydantic_field_deprecated("password", message=...)
```

If you want the compound condition (`DEFAULT_AUTHENTICATOR` *and* a non-empty password), that's a few lines inline in the existing `validate_authentication_config`. Either way I don't think it needs a new module.

**`_secret_value()` works around a type we already have.** `password` is `Optional[TransparentSecretStr]`, a real `SecretStr` at runtime — so inside the validator it's just `self.password.get_secret_value()`. The `Optional[object]` + `getattr`/`callable` version discards that, and it leaks into the tests, which assert against a hand-rolled `_SecretStrLike` and never exercise the actual field type. The empty-secret edge case is real; `get_secret_value()` covers it directly.

**We lose the `warnings.warn` half.** `add_global_warning` only reaches `pretty_print_summary`'s `click.echo` — nothing in Python's warnings system sees it. The shared helper gives you both, which also removes the need for the separate `logger.warning` in `test_connection`.

**The `report.warning` from the PR description isn't in the diff.** The body says `snowflake_v2.py` mirrors the deprecation via `self.report.warning` so it lands in the structured report/UI, but the diff only adds `logger.warning` in `test_connection`. Global warnings never reach the source report, so for UI-managed ingestion this shows up in raw stdout but not in the run's warnings — which is probably where operators would actually notice it. Worth adding for real.

**`DATAHUB_SNOWFLAKE_PASSWORD_AUTH_HARD_ERROR`** — do we need this? There's no other `*_HARD_ERROR` toggle in the repo, and Snowflake will hard-fail on its own schedule with a clearer message. If we keep it: `configuration/env_vars.py` asks that all env reads go through it, and importing `datahub.cli.env_utils` into a connector inverts the layering.

Minor: `add_global_warning` is a plain list append, so anything constructing the config more than once (Fivetran with several Snowflake destinations, the nested `connection` in `snowflake_queries`) repeats the warning. The "sources do not re-emit it, avoiding a duplicate print" comment holds only for the single-source case.

**Comment by alokr-dhub** [comment]
Thanks for the review @treff7es — all six points addressed in `ea6dafae` (test consolidation in `721fcc9`).

**1. New module.** Deleted `snowflake_auth_deprecation.py` entirely. The compound condition (`DEFAULT_AUTHENTICATOR` + non-empty password) now lives inline in the existing `validate_authentication_config` validator, and the warning text/URL moved to a module-level helper in `snowflake_connection.py`. I didn't use `pydantic_field_deprecated` because the condition isn't a simple field deprecation — it's the *combination* of auth type and a populated password — so the inline validator is the right place.

**2. `_secret_value()` workaround.** Removed. `password` is `Optional[TransparentSecretStr]` (a real `SecretStr` at runtime), so the predicate now reads `self.password.get_secret_value()` directly. The `Optional[object]` / `getattr` indirection and the `_SecretStrLike` test stand-in are gone; tests exercise the actual field type.

**3. `warnings.warn` half.** Added `warnings.warn(warning, ConfigurationWarning, stacklevel=2)` alongside `add_global_warning` in the validator. Removed the separate `logger.warning` in `test_connection` — the validator's `warnings.warn` now covers that path.

**4. `report.warning` missing.** Added `self.report.warning(get_password_auth_deprecation_warning())` in `SnowflakeV2Source.__init__` so the deprecation surfaces in the structured report / DataHub UI, not just CLI stdout.

**5. `DATAHUB_SNOWFLAKE_PASSWORD_AUTH_HARD_ERROR`.** Removed entirely. No other `*_HARD_ERROR` toggle exists in the repo, Snowflake will hard-fail on its own schedule with a clearer message, and this also drops the `datahub.cli.env_utils` import that inverted the connector → CLI layering.

**6. Duplicate-warning comment.** Removed the "sources do not re-emit it, avoiding a duplicate print" comment — it only held for the single-source case. The `add_global_warning` list-append will repeat for multi-source configs (Fivetran with several Snowflake destinations, nested `snowflake_queries`); that's an inherent property of `add_global_warning` and not something this PR tries to dedup.

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_connection.py [hunk 2]
```diff
@@ -206,8 +225,26 @@ def validate_authentication_config(self):
                     f"Should be set to 'KEY_PAIR_AUTHENTICATOR' when using key pair authentication"
                 )
 
+        # add_global_warning reaches the CLI summary; warnings.warn reaches
+        # --test-source-connection, which prints neither the summary nor the
+        # source report. SnowflakeV2Source mirrors this in its structured report.
+        if self.is_using_password_auth():
```
[the remark this answers is not in the document]
Comment (alokr-dhub) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_connection.py:231]: Stale: the `DATAHUB_SNOWFLAKE_PASSWORD_AUTH_HARD_ERROR` toggle was removed in this same commit (`ea6dafae`). The validator only emits soft warnings by design — Snowflake enforces the cutoff on its own schedule. No action.

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_v2.py [hunk 3]
```diff
@@ -161,6 +162,11 @@ def __init__(self, ctx: PipelineContext, config: SnowflakeV2Config):
         self.config: SnowflakeV2Config = config
         self.report: SnowflakeV2Report = SnowflakeV2Report()
 
+        # Mirror the deprecation in the structured report so it surfaces in the
+        # DataHub UI, not just CLI stdout.
+        if self.config.is_using_password_auth():
+            self.report.warning(get_password_auth_deprecation_warning())
```
[the remark this answers is not in the document]
Comment (alokr-dhub) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_v2.py:168]: Intentional tradeoff, not an oversight. `add_global_warning` never reaches the source's structured report (which the DataHub UI reads), and @treff7es explicitly requested `self.report.warning` for that visibility (review point #4). The two channels serve distinct audiences: `add_global_warning` for CLI summary / `--test-source-connection` (via `warnings.warn`), `self.report.warning` for the UI. The CLI double-print is the cost. Happy to collapse to one channel if you'd prefer, but the current shape is the union of what both reviewers asked for.

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_v2.py [hunk 4]
```diff
@@ -161,6 +162,11 @@ def __init__(self, ctx: PipelineContext, config: SnowflakeV2Config):
         self.config: SnowflakeV2Config = config
         self.report: SnowflakeV2Report = SnowflakeV2Report()
 
+        # Mirror the deprecation in the structured report so it surfaces in the
+        # DataHub UI, not just CLI stdout.
+        if self.config.is_using_password_auth():
```
[the remark this answers is not in the document]
Comment (alokr-dhub) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_v2.py:167]: It's a one-line wiring call to the shared helper. Per the repo's "focus on value over coverage" principles, a test asserting `report.warnings` contains the string would test the report framework's append mechanism, not this PR's logic. The predicate and `ConfigurationWarning` emission are both tested. Can add one if you want it pinned regardless.

**Comment by alokr-dhub** [comment]
Thanks for the review @treff7es — all six points addressed in `ea6dafae` (test consolidation in `721fcc9`).

**1. New module.**
> Most of `snowflake_auth_deprecation.py` already exists… I don't think it needs a new module.

Deleted `snowflake_auth_deprecation.py` entirely. The compound condition (`DEFAULT_AUTHENTICATOR` + non-empty password) now lives inline in the existing `validate_authentication_config` validator, and the warning text/URL moved to a module-level helper in `snowflake_connection.py`. I didn't use `pydantic_field_deprecated` because the condition isn't a simple field deprecation — it's the combination of auth type and a populated password — so the inline validator is the right place.

**2. `_secret_value()` workaround.**
> `password` is `Optional[TransparentSecretStr]`, a real `SecretStr` at runtime — so inside the validator it's just `self.password.get_secret_value()`… The empty-secret edge case is real; `get_secret_value()` covers it directly.

Removed. The predicate now reads `self.password.get_secret_value()` directly. The `Optional[object]` / `getattr` indirection and the `_SecretStrLike` test stand-in are gone; tests exercise the actual field type.

**3. `warnings.warn` half.**
> We lose the `warnings.warn` half. `add_global_warning` only reaches `pretty_print_summary`'s `click.echo` — nothing in Python's warnings system sees it.

Added `warnings.warn(warning, ConfigurationWarning, stacklevel=2)` alongside `add_global_warning` in the validator. Removed the separate `logger.warning` in `test_connection` — the validator's `warnings.warn` now covers that path.

**4. `report.warning` missing.**
> The `report.warning` from the PR description isn't in the diff… Global warnings never reach the source report, so for UI-managed ingestion this shows up in raw stdout but not in the run's warnings — which is probably where operators would actually notice it. Worth adding for real.

Added `self.report.warning(get_password_auth_deprecation_warning())` in `SnowflakeV2Source.__init__` so the deprecation surfaces in the structured report / DataHub UI.

**5. `DATAHUB_SNOWFLAKE_PASSWORD_AUTH_HARD_ERROR`.**
> Do we need this? There's no other `*_HARD_ERROR` toggle in the repo, and Snowflake will hard-fail on its own schedule… importing `datahub.cli.env_utils` into a connector inverts the layering.

Removed entirely. This also drops the `datahub.cli.env_utils` import that inverted the connector → CLI layering.

**6. Duplicate-warning comment.**
> `add_global_warning` is a plain list append, so anything constructing the config more than once… repeats the warning. The "sources do not re-emit it, avoiding a duplicate print" comment holds only for the single-source case.

Removed the misleading comment. The `add_global_warning` list-append will repeat for multi-source configs (Fivetran with several Snowflake destinations, nested `snowflake_queries`); that's an inherent property of `add_global_warning` and not something this PR tries to dedup.

**Review by sgomezvillamor — APPROVED** [review]
Approving to unblock — please address these before or right after merge.

**Blocker — the primary CLI channel doesn't work.** `add_global_warning()` is added during config validation (inside `Pipeline.create`), but `Pipeline.run()` calls `clear_global_warnings()` in its `finally` (`metadata-ingestion/src/datahub/ingestion/run/pipeline.py:783`), and `pretty_print_summary()` / `log_ingestion_stats()` run *after* `run()` returns (`cli/ingest_cli.py:190` → `201-202`). The list is empty by the time "Global Warnings:" prints. This is a pre-existing platform bug (came in with #19732), not introduced here, but the design rests on it working. Either fix the ordering upstream, or drop `add_global_warning` and lean on the source report + `warnings.warn`. Cursor Bugbot flagged this as "Global warning cleared before summary" and that thread is still unresolved — marked outdated by a later push, never addressed.

- Related: `warnings.warn(..., stacklevel=2)` from inside `model_validator(mode="after")` points the reported location into pydantic internals, so the stack is misleading.

**For the author to consider — only `snowflake_v2` is covered.** The diff only adds `self.report.warning(...)` to `SnowflakeV2Source.__init__`. The other consumers of `SnowflakeConnectionConfig` — `snowflake_summary.py`, `snowflake_queries.py`, `fivetran/config.py` (`SnowflakeDestinationConfig` subclasses it) and the tag propagator — have no report warning, so combined with the above all they get is `warnings.warn` on stderr. Something simple would be more than enough in those cases, e.g. for Fivetran.

**Warning — `report.warning()` contract violated.** The signature is `warning(message: LiteralString, ...)` and the docstring says the message must be a literal so identical warnings group in the report; this passes `get_password_auth_deprecation_warning()`, annotated `-> str`. mypy doesn't catch it because there are existing violations elsewhere. A module-level `Final` constant instead of the zero-arg function would satisfy `LiteralString` and read better. Also worth splitting a short `title` from the paragraph-length `message`, with the URL in `context`, as other structured warnings do.

**Nit.** The stale `snowflake_auth_deprecation.py` reference in #19694's component comment points at the module this PR deleted — worth fixing on whichever of the two merges second.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by alokr-dhub** [comment]
Thanks @sgomezvillamor @treff7es — pushed `1eb9b1f4` addressing the remaining open items:

**@sgomezvillamor:**

- **Blocker — add_global_warning cleared before summary:** Dropped `add_global_warning` entirely. `Pipeline.run()` calls `clear_global_warnings()` in its `finally` (`pipeline.py:783`), so the list is always empty by the time `pretty_print_summary()` reads it — leaning on `warnings.warn` (reaches `--test-source-connection`) + `self.report.warning` (reaches the structured report / UI) instead, as you suggested. The `add_global_warning` import is removed.
- **Warning — report.warning() LiteralString contract:** Replaced `get_password_auth_deprecation_warning()` (annotated `-> str`) with module-level `Final` constants: `SNOWFLAKE_PASSWORD_AUTH_DEPRECATION_TITLE`, `SNOWFLAKE_PASSWORD_AUTH_DEPRECATION_MESSAGE`, and `SNOWFLAKE_PASSWORD_AUTH_DEPRECATION_URL`. The `report.warning` call now passes the `Final` message literal with the URL in `context` and a short `title`, matching the pattern other structured warnings use. `get_password_auth_deprecation_warning()` is kept for the `warnings.warn` path (which takes `str`, not `LiteralString`).
- **Other consumers (Fivetran, snowflake_summary, etc.):** Acknowledged — these get `warnings.warn` from the config validator but no structured `report.warning`. I'll add a simple `report.warning` for Fivetran's `SnowflakeDestinationConfig` as a follow-up, since it doesn't have a source report in the same way.
- **stacklevel:** Kept `stacklevel=2` for now; pointing into pydantic internals is a pre-existing pattern shared with `pydantic_field_deprecated`. Happy to adjust if you'd prefer a different value.

**@treff7es:** Your earlier findings (fold into `validate_authentication_config`, drop `_secret_value()`, use `self.password.get_secret_value()` directly, add the real `report.warning` to `snowflake_v2.py`, drop `DATAHUB_SNOWFLAKE_PASSWORD_AUTH_HARD_ERROR`, dedup) were all addressed in prior commits on this branch.

**Verification:** ruff + mypy clean on both changed files; all 79 snowflake source unit tests pass.

**Review by treff7es — APPROVED** [review]
